org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLFormatter.java

```diff
@@ -656,6 +674,25 @@ private static void addAttributes(DOMNode node, XMLBuilder xmlBuilder) {
 			}
 			xmlBuilder.appendSpace();
 		}
+
+		/**
+		 * Returns true if <code>node</code> is the last non-whitespace
+		 * child of the node's owner document, and false otherwise
+		 * 
+		 * @param node the node to check
+		 * @return true if <code>node</code> is the last non-whitespace
+		 * child of the node's owner document, and false otherwise
+		 */
+		private static boolean isLastNonWhitespaceChild(DOMNode node) {

```

Comment:
The reason why I have this, is to make sure that the code doesn't add new lines after the last node. BUT this is not needed if we are going to trim the end of the document anyways.

Reply:
In other words, if `options.isTrimFinalNewlines()` is always true, we don't need this method. 

If `options.isTrimFinalNewlines()` is false, we need this method.

Reply:
@xorye if we could avoid doing that, it should be very nice, because it means that for each nodes that you wish to format, you need to loop for the DOM tree.

Today we call `this.xmlBuilder.linefeed();`  after the generate of processing instruction, prolog, etc

```java
private void formatProlog(DOMNode node) {
	addPrologToXMLBuilder(node, this.xmlBuilder);
	this.xmlBuilder.linefeed();
}
```

I wonder if it should be better to flag the xml builder to say "when you will write next content, add a new line before).

```java
private void formatProlog(DOMNode node) {
	addPrologToXMLBuilder(node, this.xmlBuilder);
	this.xmlBuilder.linefeedOnNextWrite();
}
```

and when this.xmlBuilder must be updated with new content a linefedd will be inserted?

Reply:
Understood, fixed